tests/InferTypesFuzz.elm

```diff
@@ -0,0 +1,293 @@
+
+
+stringExpr : Generator Canonical.LocatedExpr
+stringExpr =
+    Random.int 0 0x0010FFFF
+        |> Random.list 10
+        |> Random.map (List.map Char.fromCode)
+        |> Random.map String.fromList
+        |> Random.map (literal String)
+
+
+unitExpr : Generator Canonical.LocatedExpr
+unitExpr =
+    Canonical.Unit
+        |> located
+        |> Random.constant
+
+
+literal : (a -> Literal) -> a -> Canonical.LocatedExpr
+literal wrap value =
+    value
+        |> wrap
+        |> Canonical.Literal
+        |> located
+
+
+listExpr : Type -> Generator Canonical.LocatedExpr
+listExpr elementType =
+    elementType
+        |> exprGenerator
+        |> Random.list 10
+        |> Random.map Canonical.List
+        |> Random.map located
+
+
+intToIntFunctionExpr : Generator Canonical.LocatedExpr
+intToIntFunctionExpr =
+    let
+        combine argument intPart =
+            Canonical.lambda argument <|
+                located <|
+                    Canonical.Plus
+                        (located <| Canonical.Argument argument)
+                        intPart
+
+        intSubExpr =
+            Type.Int |> exprGenerator
+    in
+    Random.map2 combine
+        -- TODO: Later we will need something better to avoid shadowing.
+        randomVarName
+        intSubExpr
+        |> Random.map located
+
+
+randomVarName : Generator VarName
+randomVarName =
+    let
+        starters =
+            "abcdefghijklnmoqprstuvwxyz"
+
+        others =
+            "ABCDEFGHIJKLMNOQPRSTUVQXYZ_0123456789"
+
+        all =
+            starters ++ others
+
+        charFrom string =
+            string
+                |> String.toList
+                |> List.map Random.constant
+                |> Random.choices (Random.constant 'a')
+
+        firstChar =
+            charFrom starters
+
+        rest =
+            Random.list 5 <| charFrom all
+    in
+    Random.map2 (::) firstChar rest
+        |> Random.map String.fromList
+        |> Random.map VarName
+
+
+dumpType : Type -> String
+dumpType type_ =
+    type_
+        |> Type.toString Type.emptyState
+        |> Tuple.first
+
+
+exprShrinker : Shrinker Canonical.LocatedExpr
+exprShrinker expr =
+    case
+        expr |> Located.unwrap
+    of
+        Canonical.Literal lit ->
+            lit
+                |> shrinkLiteral
+                |> Shrink.map Canonical.Literal
+                |> Shrink.map located
+
+        Canonical.Plus left right ->
+            expr
+                |> Debug.log "shrinking (+)"
+                |> shrinkPlus left right
+
+        Canonical.List elements ->
+            elements
+                |> shrinkNonEmptyList exprShrinker
+                |> Shrink.map Canonical.List
+                |> Shrink.map located
+
+        Canonical.If { then_, else_ } ->
+            [ then_, else_ ]
+                |> List.map shrinkTo
+                |> concatShrink
+                -- Lesser hack.
+                |> shrink expr
```

Comment:
So.. the `shrink` function exists for readability reasons? So that this isn't something like
```elm
            ([ then_, else_ ]
                |> List.map shrinkTo
                |> concatShrink
            ) expr
```
?

Reply:
Exactly. It's not a great solution. I am very open to better ones.

Reply:
I think I'd rather have the code in the snippet in my original comment, and some comment explaining that the API is weird... and remove the `shrink` function. It is kinda just an additional indirection.